### PR TITLE
Expose done promise on Test

### PR DESCRIPTION
**Pull request by lgrahl**
Expose the *done* promise as `Test.done_promise` when using a `promise_test`. Allows to mix event callback handling with promise-based tests.

Real world examples: https://gist.github.com/lgrahl/de4101dc0276f28975a028a595ce05d7 and #10468

---
This change is [Reviewable](https://reviewable.io/reviews/w3c/web-platform-tests/10933)

### CODE REVIEW

File: docs/_writing-tests/testharness-api.md [hunk 1]
```diff
@@ -250,6 +250,30 @@ created Promise waiting to be fulfilled, or if the event is of a different type
...
+
+```js
+function foo() {
+  return Promise.resolve("foo");
+}
+
+promise_test(function(t) {
+  someObject.addEventListener("someevent", t.step_func(function() {
+    t.done();
+  }));
```
Comment (annevk) [inline docs/_writing-tests/testharness-api.md]: When would this be called?

Comment (lgrahl) [reply docs/_writing-tests/testharness-api.md]: I don't understand. (Maybe you can ping me in #whatwg later)

**Comment by lgrahl** [comment]
> That's a difficult requirement to document and enforce, though.

Well, I mean the `done()` method already exists, it clearly is a public method and it can be called in a `promise_test`. A suggestion for documentation:

> In a `promise_test` you have two ways to mark a test case as *done*:
>
> 1. Resolve the returned `Promise`, or
> 2. Call the `done()` method on the `Test` instance.
> 
> To use the latter it is required that you wait for the `done_promise` at the end of the test before resolving. Otherwise, the test would be marked as *done* but the returned promise could still be pending (and running).

> By awaiting the "done promise" in a test, authors are saying "this test is not done when testharness.js considers it to be done."

I would phrase it differently: "This test is not done before I explicitly mark it as done."

> From the use cases I've seen, it seems like a locally-defined promise could serve this purpose

Creating a `Promise` and waiting for it at the end would of course work but the promise and the `done()` method already exist, so it makes sense to use it for that purpose.

**Comment by jugglinmike** [comment]
> Well, I mean the `done()` method already exists, it clearly is a public method and it can be called in a `promise_test`.

I agree completely. I think that using `done` in a `promise_test` is confusing, though, because it's not clear how the two features interact and because they currently don't interact very well.

> I would phrase it differently: "This test is not done before I explicitly mark it as done."

Here's what I had in mind:

```js
promise_test(async (t) => {
  setTimeout(() => t.done());

  await t.done_promise;

  // Is the test done at this moment?
});
```

It turns out that the test *is* done at that point: the harness has already gone on to execute the next test or to report the results (if this is the final test). More troubling: any error thrown here will be ignored.

> Creating a `Promise` and waiting for it at the end would of course work but the promise and the `done()` method already exist, so it makes sense to use it for that purpose.

I'm not convinced that their existence justifies the convenience. `done` has a specific meaning which is subverted by its use in this way--that's what I tried to show in my example above.

My concern is really about using `t.done` within a `promise_test`. That's not under consideration here, but since I think that usage is problematic, I'm reluctant to endorse new features that support it.

The patch I suggested earlier reflects a different way of handling asynchronous operations. Not every API is Promise-enabled, but that doesn't have to inform the structure of the entire test (and the testing framework). Instead, you can contain the awkwardness of the callback API to the specific points that require it.

**Comment by lgrahl** [comment]
> It turns out that the test *is* done at that point: the harness has already gone on to execute the next test or to report the results (if this is the final test). More troubling: any error thrown here will be ignored.

That is a fair point. The docs would need to mention that it's okay to use promises for setup but if the *done* marking is done with the `done()` method (:smirk:) then it may not race with the promises.

However, I get your point that this is not easy to grasp. An alternative approach (very similar to your suggestion and I used it in an earlier prototype) is this:

```patch
 promise_test(async (t) => {
+  const future = new Future();
+ 
   const pc1 = new RTCPeerConnection();
   const pc2 = new RTCPeerConnection();
   t.add_cleanup(() => pc1.close());
   t.add_cleanup(() => pc2.close());

   exchangeIceCandidates(pc1, pc2);

   // Create a channel on pc1
   const dc = pc1.createDataChannel('onopen');
   dc.onopen = t.step_func(() => {
     assert_equals(dc.readyState, 'open', 'Channel should be open after onopen has fired');
-    t.done();
+    future.resolve();
   });

   await doSignalingHandshake(pc1, pc2);
-  await t.done_promise;
+  await future;
 }, 'In-band channel: Open event should be fired (local) when the data channel opens');
```

([Future](https://gist.github.com/lgrahl/1852b808273841ed806d9e0baa0aa571) definition)

File: docs/_writing-tests/testharness-api.md (hunk 1 above)
Comment (annevk) [reply docs/_writing-tests/testharness-api.md]: I mean I don't understand this example. I don't see anything dispatching `someevent`.

Comment (lgrahl) [reply docs/_writing-tests/testharness-api.md]: That's because nothing does. :slightly_smiling_face: It could be added.

**Comment by lgrahl** [comment]
So, should I use the `Future` approach instead? This is blocking #10468. :slightly_smiling_face:

**Comment by lgrahl** [comment]
Will use the `Future` since I'm getting no response.

**Pull request closed without merging**